### ISSUE TITLE
Nesting Locks for the same Event fails

### ISSUE THREAD

**Issue by julianrother** [pretix/pretix#2656]
**Describe the bug**

Attempting to acquire a lock on an event when the same thread already holds a lock on this event, causes the lock to be released too early and twice (raising a LockReleaseException). Using different instances of the event object causes a deadlock (raising LockTimeoutException) instead.

```python
@pytest.mark.django_db
def test_locking_nested_same_event_and_same_object(event):
  with event.lock():
    with event.lock():
      pass
    # Raises "pretix.base.services.locking.LockReleaseException: Lock is not
    # owned by this thread" when exiting the outer block. Exiting the inner
    # block releases the lock acquired when the outer block was entered.
    # Exiting the outer block then tries to release the lock, but fails since
    # it was already released.

@pytest.mark.django_db
def test_locking_nested_same_event_and_different_object(event):
  with event.lock():
    with scopes_disabled():
      ev = Event.objects.get(id=event.id)
      # Raises "pretix.base.services.locking.LockTimeoutException". I would
      # expect this to work, since it is still the same thread. It does not
      # work, since _lock is stored on the Event object and `event == ev`,
      # but not `event is ev`.
      with ev.lock():
        pass
```

**Expected behavior**

Judging from the locking code, the first case is supposed to work, but the code fails to keep track of whether it skipped locking due to an existing (thread-owned) lock.

The error message "Lock is not owned by this thread" suggests that lock-ownership is supposed to be tracked by the calling thread and not by the database object instance. If it would be tracked by the thread and nesting is supposed to work, the second test case should pass. Tracking locks by thread probably requires using thread-local storage.

I think nesting locks for the same event should generally work in all cases, as it is difficult to know/predict wheather a lock is held when e.g. a signal handler in a plugins is called. Alternativly signals should never be sent while a lock is held or it should be documented what signal handlers are allowed to do.

**Additional context**

In our case, we have a plugin with a signal handler for `order_paid` that changes another order of the same event with `OrderChangeManager`. `OrderChangeManager.commit` acquires a lock for the event which fails if the `order_paid` signal is sent while a lock on the same event is held. This happens (among other places) in `OrderPayment.confirm` if a payment is confirmed for an order that is expired or nearly expired (otherwise an optimization skip acquiring the lock), causing LockTimeoutExceptions.

**Comment by raphaelm** [comment]
I don't think locking is not supported in the current locking system, at least we never do it.

There is work going in #2408 to re-design the locking mechanism alltogether, but it's hard and order_paid/OrderPayment.confirm is the hardest all the situations :(

**Comment by raphaelm** [comment]
Not going to be fixed outside of #2408
